sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -60,9 +62,15 @@ case class InMemoryTableScanExec(
 
   private lazy val columnarBatchSchema = new StructType(columnIndices.map(i => relationSchema(i)))
 
+  private lazy val memoryMode = SparkEnv.get.memoryManager.tungstenMemoryMode
+
   private def createAndDecompressColumn(cachedColumnarBatch: CachedBatch): ColumnarBatch = {
     val rowCount = cachedColumnarBatch.numRows
-    val columnVectors = OnHeapColumnVector.allocateColumns(rowCount, columnarBatchSchema)
+    val columnVectors = if (memoryMode == MemoryMode.ON_HEAP) {
+      OnHeapColumnVector.allocateColumns(rowCount, columnarBatchSchema)
+    } else {
+      OffHeapColumnVector.allocateColumns(rowCount, columnarBatchSchema)
```

Comment:
where do we close it?

Reply:
Great catch, we need to use `context.addTaskCompletionListener` for close this.

Reply:
For now, I added `context.addTaskCompletionListener`, but I did not add `batch.close` in generated code.
This is because ParquetReader reuses a `ColumnarBatch. Thus, to call `batch.close` causes runtime exception during the reuse.